src/esm.ts

```diff
@@ -147,12 +149,40 @@ export function createEsmHooks(tsNodeService: Service) {
       return defer();
     }
 
-    // pathname is the path to be resolved
+    let candidateSpecifiers: string[] = [specifier];
+
+    if (context.parentURL) {
+      const parentUrl = new URL(context.parentURL);
+      if (parentUrl.pathname && extname(parentUrl.pathname) === '.ts') {
+        const mappedSpecifiers = mapPath(specifier);
+        if (mappedSpecifiers) {
+          candidateSpecifiers = mappedSpecifiers.map((path) =>
+            pathToFileURL(path).toString()
+          );
+        }
+      }
+    }
+
+    let candidateSpecifier: string | undefined;
+    while ((candidateSpecifier = candidateSpecifiers.shift())) {
+      try {
+        return await nodeResolveImplementation.defaultResolve(
+          candidateSpecifier,
+          context,
+          defaultResolve
+        );
+      } catch (err) {
+        const isNotFoundError = (<any>err).code === 'ERR_MODULE_NOT_FOUND';
+        if (isNotFoundError && candidateSpecifiers.length > 0) {
+          continue;
+        } else {
+          throw err;
```

Comment:
I wonder, should we generate a new kind of error indicating that we tried many candidates?  `err` talks specifically about the *last* candidate which was tried.  Will that be confusing to the user?

In the following example, assume `config` maps to both `src/config` and `src/codegen/config`.
The user has a typo in their specifier, but the error makes them think it has skipped over the path mapping to `src/config`:

```
$ node
> import('config/defalts')
Uncaught:
Error [ERR_MODULE_NOT_FOUND]: Cannot find package '/d/ts-node/path-mappings/src/codegen/config/defalts' imported from /d/ts-node/path-mapping/repl
```

An example of the style of message we might return instead:

```
Uncaught:
Error [ERR_MODULE_NOT_FOUND]: Cannot find module for specifier 'config/defalts' imported from /d/ts-node/path-mapping/repl using TypeScript path mappings
Candidate specifiers attempted:
- /d/ts-node/path-mappings/src/config/defalts
- /d/ts-node/path-mappings/src/codegen/config/defalts
```

This error makes it clear that `src/config` was attempted.

Reply:
> I wonder, should we generate a new kind of error indicating that we tried many candidates? err talks specifically about the last candidate which was tried. Will that be confusing to the user?

That’s a good point and I like your suggestion of providing more information. I implemented this in https://github.com/TypeStrong/ts-node/pull/1585/commits/a19454b107bf824a4324824251545595ee7d7d9b